### DISCUSSION TITLE
challenge configuring under DOOM Emacs

### DISCUSSION THREAD

**Discussion by jeffkowalski** [kickingvegas/casual-calc#99] [Q&A] [answered]
Using emacs 29.3, Doom core v3.0.0-pre, Doom modules v24.03.0-pre

I tangle this in my config:
```
#+BEGIN_SRC emacs-lisp :tangle packages.el
(package! casual
  :recipe (:host github :repo "kickingvegas/Casual"))
#+END_SRC
#+BEGIN_SRC emacs-lisp
(use-package! casual
  :bind (:map calc-mode-map
         ("C-o" . casual-main-menu)))
#+END_SRC
```

Which, upon ```doom sync``` results in the following error:
```
~ λ doom sync
> Tangling your literate config...
  ✓ Done tangling 2 file(s)!
- Using Emacs 29.3 @ /usr/bin/emacs
> Synchronizing "default" profile...
  x There was an unexpected runtime error
    Message: Symbol's function definition is void
    Details: (transient-prefix)
    Backtrace:
      (transient-prefix :command 'casual-main-menu)
      (put 'casual-main-menu 'transient--prefix (transient-prefix :command 'casual-main-menu))
      (load-with-code-conversion "/home/jeff/.config/emacs/.local/straight/build-29.3/casual/casual-autoloads.el" "/home/jeff/.config/emacs/.l...
      (load "/home/jeff/.config/emacs/.local/straight/build-29.3/casual/casual-autoloads.el" nil nomessage)
      (progn (load autoloads-file nil 'nomessage))
      (if (file-exists-p autoloads-file) (progn (load autoloads-file nil 'nomessage)))
      (let ((autoloads-file (straight--autoloads-file package))) (if (file-exists-p autoloads-file) (progn (load autoloads-file nil 'nomessage...
      (straight--load-package-autoloads "casual")
      (let ((name (symbol-name package))) (add-to-list 'load-path (directory-file-name (straight--build-dir name))) (straight--load-package-au...
      (let ((package (car tail))) (straight-register-package package) (let ((name (symbol-name package))) (add-to-list 'load-path (directory-f...
      (while tail (let ((package (car tail))) (straight-register-package package) (let ((name (symbol-name package))) (add-to-list 'load-path ...
      (let ((tail (cl-delete-duplicates packages :test #'equal))) (while tail (let ((package (car tail))) (straight-register-package package) ...
    ! Wrote extended backtrace to ~/.config/emacs/.local/state/logs/cli.doom.240331200736.210476.error
  ! Script was abruptly aborted, leaving Doom in an incomplete state!
  - Run 'doom sync' to repair it.
  ✓ Finished in 1.57351s
```

**Comment by kickingvegas** [comment]
@jeffkowalski Thanks for the input. Unfortunately I'm not familiar with DOOM Emacs, much less using a literate configuration via Org. Will convert this to a discussion item so that perhaps another DOOM user can provide guidance.

I noticed that in your `use-package` statement that did not include `:ensure t`. Perhaps that can help?

**Comment by jeffkowalski** [comment]
That's a good guess, @kickingvegas , but ```:ensure t``` is ignored in doom emacs, in preference for using the ```package!``` command.  What's odd is that the error is occurring when casual is first referenced in ```package!```, rather than in ```use-package!```.  If I comment out the ```use-package!``` I still get the error on startup. 
In fact, this is all that's required to create the error above:
```
#+BEGIN_SRC emacs-lisp :tangle packages.el
(package! casual)
#+END_SRC
```
This simply puts ```(package! casual)``` into ```packages.el```.  A subsequent ```doom sync``` on the commandline results in the error above.

 It feels to me (without real knowledge here), that something is depending on transient earlier than it should, earlier than it is in fact loaded.

**Comment by kickingvegas** [comment]
@jeffkowalski that's a bummer. Do you use magit? That should pull in Transient. Also, open invite for other DOOM users to chime in.

**Comment by jeffkowalski** [comment]
yes, I use magit, but it doesn't have the issue of trying to reference transient before runtime.
Interestingly, if I remove the autoload here https://github.com/kickingvegas/Casual/blob/ace04facf96e71b097d3ba83f64b1963af4b8f13/lisp/casual.el#L72
Then everything works fine.  (By "everything", I mean the lines I described in my literate config create the changes in the respective files, ```doom sync``` runs without issue to pull the package and compile it, emacs starts without error, ```M-x calc``` works, and my ```C-o``` pulls up the counsel main menu.)
Is that autoload strictly necessary?

**Comment by kickingvegas** [comment]
@jeffkowalski getting guidance from the DOOM Discord. Try this statement:

```
;;;###autoload (autoload 'casual-main-menu "casual" nil t)
```

**Reply by jeffkowalski** [reply]
Yes, 
```
;;;###autoload (autoload 'casual-main-menu "casual" nil t)
```
appears to work.  Thanks for taking this up on Discord.  I hope this is a useful solution.

**Comment by kickingvegas** [comment]
Ok will make a patch with this change.

**Comment by kickingvegas** [comment, marked as answer]
Casual v1.3.1 update with explicit autoload now on MELPA. https://melpa.org/#/casual @jeffkowalski please verify this works with DOOM and if so, then mark as answered this discussion thread.

Thanks!

**Reply by jeffkowalski** [reply]
I've confirmed that the version v1.3.1 now available through MELPA contains the appropriate fix.
